content/filterx/_index.md

```diff
@@ -0,0 +1,377 @@
+
+```shell
+log {
+    source(s1);
+    filterx demo_filterx {
+        ${HOST} == "example";
+        ${MESSAGE} =~ "deny";
+    };
+    destination(d1);
+};
+```
+
+You can use filterx blocks together with other blocks in a log path, for example, use a parser before/after the filterx block if needed.
+
+<!-- FIXME what is mutable/immutable writable/read-only > devs to write a draft  -->
+
+## Filterx statements
+
+A filterx block contains one or more filterx statements. The order of the statements is important, as they are sequentially processed.  If any of the statements is false (or results in an error), {{< product >}} drops the message from that log path.
+
+Filterx statements can be one of the following:
+
+- A comparison, for example, `$HOST == "my-host";`. This statement is true only for messages where the `$HOST` field is `my-host`. Such simple comparison statements can be the equivalents of [traditional filter functions]({{< relref "/chapter-routing-filters/filters/reference-filters/_index.md" >}}).
+- A value assignment for a [name-value pair or a local variable](#variable-scope), for example, `$my-field = "bar";`. The left-side variable automatically gets the type of the right-hand expression.
+- A conditional statement ( `if (expr) { ... } else { ... };`) that allows you evaluate complex decision trees.
+- A declaration of a [pipeline variable](#variable-scope), for example, `declare my-pipeline-variable = "something";`.
+
+{{% alert title="Note" color="info" %}}
+- The `true;` and `false;` strings are also valid statements. They can be useful in complex conditional (if/elif/else) statements.
+- A name-value pair or a variable in itself is also a statement. For example, `$HOST;`. If the name-value pair or variable is empty or doesn't exist, the statement is false.
+{{% /alert %}}
+
+When you assign the value of a variable using another variable (for example, `$MESSAGE = "$HOST"`), {{< product >}} copies the current value of the `$HOST` variable. If a statement later changes the value of the `$HOST` field, the `$MESSAGE` field won't change. For example:
+
+```shell
+filterx {
+  ${HOST} = "first-hostname";
+  ${MESSAGE} = ${HOST}; # The value of $MESSAGE is first-hostname
+  ${HOST} = "second-hostname"; # The value of $MESSAGE is still first-hostname
+};
+```
+
+The same is true for complex objects, like JSON, for example:
+
+```shell
+js = json_object({
+    "key": "value",
+    "second-key": "another-value"
+});
+
+${MESSAGE} = js;
+
+js.third_key = "third-value-not-available-in-$MESSAGE";
+```
+
+## Data model and scope {#scoping}
+
+Each filterx block can access data from the following elements.
+
+- Macros and name-value pairs of the message being processed (for example, `$PROGRAM`). The names of macros and name-value pairs begin with the `$` character. If you define a new variable in a filterx block and its name begins with the `$` character, it's automatically added to the name-value pairs of the message.
```

Comment:
if that's the case (I've seen examples of it in this doc), I would probably mention here that `$foo` can be written as `${foo}`

Reply:
and it indeed can. the reason braces are permitted as the names can contain stuff like dots and other special characters.

${.SDATA.foo.bar} works for instance